gossip/src/crds_value.rs

```diff
@@ -494,80 +499,125 @@ impl Sanitize for NodeInstance {
+    pub compressed_slots: Vec<u8>,
+    pub uncompressed_bytes: u64,
+    pub last_voted_slot: Slot,
     pub last_voted_hash: Hash,
     pub shred_version: u16,
 }
 
 impl Sanitize for RestartLastVotedForkSlots {
     fn sanitize(&self) -> std::result::Result<(), SanitizeError> {
-        if self.slots.is_empty() {
+        if self.last_voted_slot == 0 {
             return Err(SanitizeError::InvalidValue);
         }
-        self.slots.sanitize()?;
         self.last_voted_hash.sanitize()
     }
 }
 
 impl RestartLastVotedForkSlots {
-    pub fn new(from: Pubkey, now: u64, last_voted_hash: Hash, shred_version: u16) -> Self {
-        Self {
+    pub fn new(
+        from: Pubkey,
+        now: u64,
+        last_voted_fork: &mut [Slot],
+        last_voted_hash: Hash,
+        shred_version: u16,
+    ) -> Result<Self, String> {
+        let last_voted_slot;
+        let mut compressed_slots;
+        let uncompressed_bytes;
+        if last_voted_fork.is_empty() {
+            return Err("Last voted slot must be specified".to_string());
+        } else {
+            last_voted_fork.sort_by_key(|a| Reverse(*a));
+            last_voted_slot = last_voted_fork[0];
+            let max_size = last_voted_slot.saturating_sub(*last_voted_fork.last().unwrap()) + 1;
+            let mut uncompressed_bitvec = BitVec::new_fill(false, max_size);
+            for slot in last_voted_fork {
+                uncompressed_bitvec.set(last_voted_slot - *slot, true);
+            }
+            compressed_slots = Vec::with_capacity(MAX_RESTART_LAST_VOTED_FORK_SLOTS_SPACE);
+            let mut compressor = Compress::new(Compression::best(), false);
+            // If there is more input than we can fit into given output, compress_vec() will return Ok
+            // but will not terminate the output buf correctly for decompress, so we need to start all
+            // over again with shorter input so compress can actually finish.
+            match compressor.compress_vec(
+                &uncompressed_bitvec.clone().into_boxed_slice(),
+                &mut compressed_slots,
+                FlushCompress::Finish,
+            ) {
+                Ok(flate2::Status::Ok) => {
+                    let in_bytes = compressor.total_in();
+                    uncompressed_bitvec.truncate((in_bytes - 1) * 8);
+                    compressor = Compress::new(Compression::best(), false);
+                    if let Err(e) = compressor.compress_vec(
+                        &uncompressed_bitvec.into_boxed_slice(),
+                        &mut compressed_slots,
+                        FlushCompress::Finish,
+                    ) {
+                        return Err(e.to_string());
+                    }
+                }
+                Err(e) => return Err(e.to_string()),
+                // compression ended successfully, proceed.
+                Ok(flate2::Status::StreamEnd) => (),
+                Ok(flate2::Status::BufError) => (),
```

Comment:
I am guessing in this case some of the slots are dropped because there is no more buffer to write to.
Can you comment this out here in the code?
Also can you have a test that in this case the other nodes will still receive a valid payload and can decompress to a valid list of slots consistent with the most recent slots in the original vector.